### PR TITLE
refactor AUC OP and add its CUDA Kernel

### PR DESCRIPTION
**Pull request by hutuxian**
Optimize the time complexity of AUC OP from O(slide_step * num_bucket) to O(num_bucket).
Add its CUDA Kernel. And the auc kernel will be optimized from O(num_bucket) to O(log(num_bucket)), which is a future work.

### CODE REVIEW

File: python/paddle/fluid/layers/metric_op.py [hunk 1]
```diff
@@ -170,17 +170,17 @@ def auc(input,
     batch_stat_pos = helper.create_global_variable(
         persistable=True,
         dtype='int64',
-        shape=[slide_steps, num_thresholds + 1])
+        shape=[(1 + slide_steps) * (num_thresholds + 1) + 1])
```
Comment (gavin1332) [inline python/paddle/fluid/layers/metric_op.py:178]: add comments for data layout description

File: paddle/fluid/operators/metrics/auc_op.cu [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+using LoDTensor = framework::LoDTensor;
+#define CUDA_KERNEL_LOOP(i, n)                                 \
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < (n); \
+       i += blockDim.x * gridDim.x)
+
+__global__ void ClearObsoleteDataKernel(int64_t *pos, int64_t *neg,
+                                        const int bucket_length,
+                                        const int slide_steps) {
+  int cur_step_index =
+      static_cast<int>(pos[(slide_steps + 1) * bucket_length] + 0.5) %
```
Comment (gavin1332) [inline paddle/fluid/operators/metrics/auc_op.cu]: useless 0.5

File: paddle/fluid/operators/metrics/auc_op.cu [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+  int cur_step_begin = 0;
+  if (slide_steps > 0) {
+    int cur_step_index =
+        static_cast<int>(pos[(slide_steps + 1) * (1 + num_thresholds)] + 0.5) %
+        slide_steps;
+    cur_step_begin = cur_step_index * (1 + num_thresholds);
+  }
+  CUDA_KERNEL_LOOP(i, numel) {
+    auto predict_data = pred_data[i * inference_width + (inference_width - 1)];
+    // cannot pass the compilation when using PADDLE_ENFORCE_LE
```
Comment (gavin1332) [inline paddle/fluid/operators/metrics/auc_op.cu]: HOW to use enforce statements in cuda kernel

File: paddle/fluid/operators/metrics/auc_op.cu [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+template <typename T>
+__global__ void AddDataKernel(const int64_t *label_data, const T *pred_data,
+                              const int inference_width,
+                              const int num_thresholds, int64_t *pos,
+                              int64_t *neg, const int numel,
+                              const int slide_steps) {
+  int cur_step_begin = 0;
+  if (slide_steps > 0) {
+    int cur_step_index =
+        static_cast<int>(pos[(slide_steps + 1) * (1 + num_thresholds)] + 0.5) %
```
Comment (gavin1332) [inline paddle/fluid/operators/metrics/auc_op.cu]: useless 0.5

File: paddle/fluid/operators/metrics/auc_op.cu [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    neg[sum_step_begin + i] -= neg[cur_step_begin + i];
+    pos[cur_step_begin + i] = neg[cur_step_begin + i] = 0;
+  }
+}
+
+__global__ void UpdateSumDataKernel(int64_t *pos, int64_t *neg,
+                                    const int bucket_length,
+                                    const int slide_steps) {
+  int cur_step_index =
+      static_cast<int>(pos[(slide_steps + 1) * bucket_length] + 0.5) %
```
Comment (gavin1332) [inline paddle/fluid/operators/metrics/auc_op.cu]: useless 0.5

File: paddle/fluid/operators/metrics/auc_op.cu [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+      return;
+    }
+    // the last number of origin_stat_pos store the index should be used in
+    // current step
+    int cur_step_index =
+        static_cast<int>(origin_stat_pos[(slide_steps + 1) * bucket_length]) % slide_steps;
+    int cur_step_begin = cur_step_index * bucket_length;
+    int sum_step_begin = slide_steps * bucket_length;
+
+    ClearObsoleteDataKernel<<<(bucket_length + PADDLE_CUDA_NUM_THREADS - 1) /
```
Comment (gavin1332) [inline paddle/fluid/operators/metrics/auc_op.cu:196]: could the these three CUDA KERNELs be merged?

**Review by gavin1332 — COMMENTED** [review]
could be appear in release 1.7 as a function enforcement.

File: paddle/fluid/operators/metrics/auc_op.cu (hunk 3 above)
Comment (hutuxian) [reply paddle/fluid/operators/metrics/auc_op.cu]: Confirmed with LuoTao, just use PADDLE_ENFORCE

**Review by gavin1332 — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LGTM for the paddle_enforce in cuda kernel, @chenwhql will update the wiki later.

**Pull request merged**